feat(typing): better typing

mehdigmira opened the pull request:
Better typing, so that mypy would emit errors when messages are built with the wrong parameters, and that results types are correctly inferred. More details below. PEP 692 and PEP 646 will allow even better typing.

```
@remoulade.actor(store_results=True)  # only callable actors are well handled, from a typing point of view
def request(uri: str) -> str:
    ...

@remoulade.actor(store_results=True)  # well handled
def add(x: int, y: int) -> int:
    ...

class SumTask(GenericActor):  # not well handled
    def perform(self, x: int, y: int) -> int:
        ...

add.message(1, 2)  # no error
add.message(x=1, y=2)  # no error
add.message(x=1, y=2, z=1)  # mypy error
add.message(1, 2, 3)  # mypy error

reveal_type(add.send(1, 2).result.get())  #  Union[builtins.int, remoulade.results.errors.ErrorStored]
reveal_type(add.send(1, 2).result.get(raise_on_error=True))  # builtins.int

reveal_type(
    group(
        [
            request.message_with_options(args=[1]),
            request.message_with_options(args=["x"]),
        ]
    ).results.get()
)  # typing.Iterator[Union[remoulade.results.errors.ErrorStored, builtins.str]]
reveal_type(
    group(
        [
            request.message_with_options(args=[1]),
            request.message_with_options(args=["x"]),
        ]
    ).results.get(raise_on_error=True)
)  # typing.Iterator[builtins.str]

reveal_type(
    group(
        [
            request.message_with_options(args=[1]),
            add.message(1, 2),
            request.message_with_options(args=["x"]),
        ]
    ).results.get()
)  # typing.Iterator[builtins.object]
```

antoinerabany commented on remoulade/collection_results.py:
```diff
@@ -94,9 +99,35 @@ def completed_count(self) -> int:
+    ) -> Iterator[Union[ErrorStored, ResultT]]:
+        ...
+
+    @overload
```
why this one ?

mehdigmira replied:
This allows inferring group().get() type

antoinerabany commented on tests/state/test_state_api.py:
```diff
@@ -233,7 +233,7 @@ def test_requeue_message(self, stub_broker, do_work, api_client, state_middlewar
+        message: Message[int] = Message(
             queue_name="default",
             actor_name="do_work",
             args=(),
```
please update the changelog file

mehdigmira replied:
Updated this

mehdigmira commented:
@antoinerabany need to update .github/workflows/lint.yml, so that mypy command is `mypy --install-types --non-interactive --show-traceback --enable-incomplete-features --namespace-packages`, but i don't have the right access rights

```
refusing to allow a Personal Access Token to create or update workflow `.github/workflows/lint.yml` without `workflow` scope)
```

mehdigmira commented:
@antoinerabany this requires a new review. I've reworked the PR and added tests + typed `options`. Tests fail though, with a connection error to the postgres database, do you have a clue what would cause that ? I didn't really change that part of the code. Those tests pass locally.

antoinerabany commented on remoulade/composition.py:
```diff
@@ -222,29 +253,29 @@ def __len__(self) -> int:
         composition_id = options.get("composition_id", self.group_id)
-        cancel_on_error = options.get("composition_id", self.cancel_on_error)
-        options = {
+        cancel_on_error = options.get("cancel_on_error", self.cancel_on_error)
```
yes

antoinerabany commented:
You are admin now, there shouldn't be porblems anymore

mehdigmira commented:
Closing in favor of https://github.com/wiremind/remoulade/pull/338
